# Data Loading

In [1]:
import pandas as pd

room_attributes = pd.read_csv('/kaggle/input/hotel-recommendation/Hotel_Room_attributes.csv')
hotel_details = pd.read_csv('/kaggle/input/hotel-recommendation/Hotel_details.csv')

print('Jumlah data atribut kamar hotel:', len(room_attributes))
print('Jumlah data hotel:', len(hotel_details['hotelid'].unique()))


Jumlah data atribut kamar hotel: 165873
Jumlah data hotel: 106193


# Univariate Exploratory Data Analysis

In [2]:
room_attributes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 165873 entries, 0 to 165872
Data columns (total 5 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   id               165873 non-null  int64 
 1   hotelcode        165873 non-null  int64 
 2   roomamenities    161054 non-null  object
 3   roomtype         165873 non-null  object
 4   ratedescription  161054 non-null  object
dtypes: int64(2), object(3)
memory usage: 6.3+ MB


In [3]:
room_attributes.head()

id  hotelcode                                      roomamenities  \
0  50677497     634876  Air conditioning: ;Alarm clock: ;Carpeting: ;C...   
1  50672149    8328096  Air conditioning: ;Closet: ;Fireplace: ;Free W...   
2  50643430    8323442  Air conditioning: ;Closet: ;Dishwasher: ;Firep...   
3  50650317       7975  Air conditioning: ;Clothes rack: ;Coffee/tea m...   
4  50650318       7975  Air conditioning: ;Clothes rack: ;Coffee/tea m...   

               roomtype                                    ratedescription  
0           Double Room       Room size: 15 m²/161 ft², Shower, 1 king bed  
1         Vacation Home  Shower, Kitchenette, 2 bedrooms, 1 double bed ...  
2         Vacation Home  Shower, Kitchenette, 2 bedrooms, 1 double bed ...  
3  Standard Triple Room    Room size: 20 m²/215 ft², Shower, 3 single beds  
4  Standard Triple Room    Room size: 20 m²/215 ft², Shower, 3 single beds

### Fitur pada Dataset Hotel Room Atribut  
1. **id** (`int64`)  
   ID unik untuk setiap kamar yang tersedia dalam dataset.  

2. **hotelcode** (`int64`)  
   Kode unik untuk mengidentifikasi hotel tempat kamar tersebut berada.  

3. **roomamenities** (`object`)  
   Daftar fasilitas atau kelengkapan kamar, seperti AC, Wi-Fi, dan TV.  

4. **roomtype** (`object`)  
   Jenis kamar yang tersedia, seperti **Standard Room, Deluxe Room, Suite**, dll.  

5. **ratedescription** (`object`)  
   Deskripsi rating kamar berdasarkan kategori tertentu.  

In [4]:
hotel_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108048 entries, 0 to 108047
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            108048 non-null  int64  
 1   hotelid       108048 non-null  int64  
 2   hotelname     108048 non-null  object 
 3   address       102955 non-null  object 
 4   city          108047 non-null  object 
 5   country       108048 non-null  object 
 6   zipcode       83486 non-null   float64
 7   propertytype  108048 non-null  object 
 8   starrating    108048 non-null  int64  
 9   latitude      108048 non-null  float64
 10  longitude     108048 non-null  float64
 11  Source        108048 non-null  int64  
 12  url           107937 non-null  object 
 13  curr          108048 non-null  object 
dtypes: float64(3), int64(4), object(7)
memory usage: 11.5+ MB


In [5]:
hotel_details.head()

/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.10/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in greater
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()


id  hotelid                 hotelname                 address     city  \
0  46406  1771651  Mediteran Bungalow Galeb            Vukovarska 7     Omis   
1  46407   177167             Hotel Polonia        Plac Teatralny 5    Torun   
2  46408  1771675         Rifugio Sass Bece  Belvedere del Pordoi,1  Canazei   
3  46409   177168            Madalena Hotel                 Mykonos  Mykonos   
4  46410  1771718         Pension Morenfeld  Mair im Korn Strasse 2  Lagundo   

   country  zipcode   propertytype  starrating   latitude  longitude  Source  \
0  Croatia  21310.0  Holiday parks           4  43.440124  16.682505       2   
1   Poland      NaN         Hotels           3  53.012329  18.603800       5   
2    Italy  38032.0         Hotels           3  46.477920  11.813350       2   
3   Greece  84600.0         Hotels           3  37.452316  25.329849       5   
4    Italy  39022.0         Hotels           3  46.682780  11.131736       2   

                                                 url curr  
0  https://www.booking.com/hotel/hr/bungalow-luxu...  EUR  
1  https://www.agoda.com/en-gb/hotel-polonia/hote...  EUR  
2  http://www.booking.com/hotel/it/rifugio-sass-b...  EUR  
3  https://www.agoda.com/en-gb/madalena-hotel/hot...  EUR  
4  http://www.booking.com/hotel/it/pension-morenf...  EUR

### Fitur pada Dataset Hotel Details
1. **id** (`int64`)  
   ID unik untuk setiap hotel dalam dataset.  

2. **hotelcode** (`int64`)  
   Kode unik hotel yang dapat digunakan untuk menghubungkan dataset kamar dan hotel.  

3. **hotelname** (`object`)  
   Nama hotel, seperti **Grand Hotel, Sunset Resort, dll.**  

4. **address** (`object`)  
   Alamat lengkap hotel, yang dapat mencakup **jalan, nomor, dan kawasan**.  

5. **city** (`object`)  
    Kota tempat hotel tersebut berada, seperti **Jakarta, Bali, Tokyo, dll.**  

6. **country** (`object`)  
    Negara tempat hotel berada, seperti **Indonesia, Jepang, Amerika Serikat, dll.**  

7. **zipcode** (`float64`)  
    Kode pos lokasi hotel.  

8. **propertytype** (`object`)  
    Jenis properti hotel, seperti **Hotel, Guesthouse, Hostel, Villa**, dll.  

9. **starrating** (`int64`)  
    Rating bintang hotel (misalnya **3, 4, atau 5 bintang**).  

10. **latitude** (`float64`)  
    Koordinat **lintang** lokasi hotel dalam derajat desimal.  
    - **Nilai positif** → Belahan bumi utara (N).  
    - **Nilai negatif** → Belahan bumi selatan (S).  

11. **longitude** (`float64`)  
    Koordinat **bujur** lokasi hotel dalam derajat desimal.  
    - **Nilai positif** → Belahan bumi timur (E).  
    - **Nilai negatif** → Belahan bumi barat (W).  

12. **Source** (`int64`)  
    Sumber data hotel, yang mungkin digunakan untuk tracking data dari berbagai platform.  

13. **url** (`object`)  
    URL halaman pemesanan hotel atau informasi resmi.  

14. **curr** (`object`)  
    Mata uang yang digunakan untuk harga kamar hotel, seperti **IDR, USD, EUR**, dll.  

# Data Preprocessing

In [6]:
hotel_details.rename(columns = {'hotelid':'hotelcode'},inplace = True)

Mengubah variabel `hotelid` di dataset hotel_details menjadi `hotelcode` agar seragam dengan diberi nama df_merged

In [7]:
df_merged = room_attributes.merge(hotel_details, on='hotelcode', how='left')

Menggabungkan dataset room_attributes dengan hotel_details berdasarkan `hotelcode`

In [8]:
df_merged.shape

(181415, 18)

Setelah digabungkan (merge) jumlah dataset menjadi 181415 sampel dengan 18 variabel/fitur

In [9]:
df_cleaned = df_merged[[
    'hotelname', 'roomtype',
    'city', 'country'
]]

Mengambil variabel 'hotelname', 'roomtype', 'city', 'country', yang akan digunakan untuk membangun sistem rekomendasi

In [10]:
df_cleaned.head()

hotelname              roomtype           city         country
0  The Old Cider House           Double Room  Nether Stowey  United Kingdom
1            Pantlleni         Vacation Home     Beddgelert  United Kingdom
2       Willow Cottage         Vacation Home       Beverley  United Kingdom
3  Apollo Hotel London  Standard Triple Room         London  United Kingdom
4  Apollo Hotel London  Standard Triple Room         London  United Kingdom

# Data Preparation

## Check missing values

In [11]:
df_cleaned.isnull().sum()

hotelname    0
roomtype     0
city         0
country      0
dtype: int64

In [14]:
df.shape

(181415, 4)

## Check Duplicated Values

In [15]:
df.duplicated().sum()

131919

ada 131919 nilai yang terduplikasi

## Handling Duplicated Values

In [16]:
df.drop_duplicates(inplace = True)
df.shape

(49496, 4)

menghapus data yang terduplikasi. Sehingga dataset berkurang menjadi 49496

## Cleaning and Concat Dataset values

In [17]:
import re

def clean_text(text):
    text = re.sub(r'[:;]', '', text)  # Hapus karakter ": ;"
    text = text.lower().strip()  # Ubah ke lowercase dan hapus spasi
    return text

# Cleaning setiap kolom
df['hotelname'] = df['hotelname'].apply(clean_text)
df['roomtype'] = df['roomtype'].apply(clean_text)
df['city'] = df['city'].apply(clean_text)
df['country'] = df['country'].apply(clean_text)

df['tags'] = df['hotelname'] + ' ' + df['roomtype'] + ' ' + df['city'] + ' ' + df['country']


semua fitur menjadi satu yaitu fitur `tags`. Fitur `tags` akan menjadi fitur utama yang selanjutnya diolah untuk menjadi dasar sistem rekomendasi hotel.

In [18]:
df.head()

hotelname              roomtype           city         country  \
0  the old cider house           double room  nether stowey  united kingdom   
1            pantlleni         vacation home     beddgelert  united kingdom   
2       willow cottage         vacation home       beverley  united kingdom   
3  apollo hotel london  standard triple room         london  united kingdom   
6  apollo hotel london    deluxe triple room         london  united kingdom   

                                                tags  
0  the old cider house double room nether stowey ...  
1  pantlleni vacation home beddgelert united kingdom  
2  willow cottage vacation home beverley united k...  
3  apollo hotel london standard triple room londo...  
6  apollo hotel london deluxe triple room london ...

# Modeling

menggunakan Content-Based Filtering untuk mencari similarity antara hotel, tipe kamar dan dan lokasi hotel

## Feature Extraction

In [19]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD

tfidf = TfidfVectorizer(max_features=1000, stop_words='english')

tfidf_matrix = tfidf.fit_transform(df['tags'])

svd = TruncatedSVD(n_components=1000)
reduced_matrix = svd.fit_transform(tfidf_matrix)

- TF-IDF mengubah teks menjadi vektor numerik berbobot berdasarkan frekuensi kata.
- Truncated SVD (LSA) mereduksi dimensi agar lebih efisien tanpa kehilangan banyak informasi.

## Content-Based Filtering with Cosine Similarity

In [20]:
from sklearn.metrics.pairwise import cosine_similarity
 
cosine_sim = cosine_similarity(reduced_matrix) 
cosine_sim

array([[1.        , 0.12090303, 0.09691459, ..., 0.13084286, 0.12275963,
        0.07803286],
       [0.12090303, 1.        , 0.80158945, ..., 0.07540598, 0.07074754,
        0.07674847],
       [0.09691459, 0.80158945, 1.        , ..., 0.06044464, 0.05671048,
        0.06152077],
       ...,
       [0.13084286, 0.07540598, 0.06044464, ..., 1.        , 0.88108588,
        0.92143128],
       [0.12275963, 0.07074754, 0.05671048, ..., 0.88108588, 1.        ,
        0.86450694],
       [0.07803286, 0.07674847, 0.06152077, ..., 0.92143128, 0.86450694,
        1.        ]])

cosine similarity digunakan untuk menghitung kesamaan antar data berdasarkan matriks fitur yang telah direduksi.

## Recomendation Function

In [21]:
def recommend(roomtype, country, city, num_recommendations=10):
    # Normalisasi input untuk pencarian
    roomtype, country, city = map(str.lower, [roomtype, country, city])
    
    # Filter data berdasarkan input pengguna
    filtered_df = df[
        (df['roomtype'].str.lower() == roomtype) & 
        (df['country'].str.lower() == country) & 
        (df['city'].str.lower() == city) 
    ]
    
    if filtered_df.empty:
        return "Tidak ada hotel yang sesuai dengan filter yang diberikan."
    
    # Ambil index dari hotel yang sesuai
    idx_list = filtered_df.index.tolist()
    recommended_hotels = []

    for idx in idx_list:
        try:
            idx_in_similarity = df.index.get_loc(idx) 
            similar_hotels = sorted(
                list(enumerate(cosine_sim[idx_in_similarity])),
                key=lambda x: x[1], reverse=True
            )[1:num_recommendations+1]
            
            # Tambahkan ke hasil rekomendasi
            for hotel_idx, similarity_score in similar_hotels:
                recommended_hotels.append({
                    "hotelname": df.iloc[hotel_idx]['hotelname'],
                    "roomtype": df.iloc[hotel_idx]['roomtype'],
                    "country": df.iloc[hotel_idx]['country'],
                    "city": df.iloc[hotel_idx]['city'],
                    "similarity": round(similarity_score, 2)
                })

        except IndexError:
            continue  # Lewati jika indeks tidak ditemukan
    
    # Konversi ke DataFrame
    result_df = pd.DataFrame(recommended_hotels).drop_duplicates().head(num_recommendations)
    
    # Jika tidak ada rekomendasi yang ditemukan berdasarkan kemiripan
    if result_df.empty:
        return "Tidak ada rekomendasi yang ditemukan berdasarkan kemiripan."

    return result_df


Menggunakan cosine similarity untuk membangun fungsi rekomendasi berdasarkan tipe kamar, negara, dan kota

# Evaluation

In [22]:
recommend('double room', 'Italy', 'Venice')

hotelname              roomtype country    city  similarity
0  hotel antica fenice           double room   italy  venice        1.00
1       hotel dolomiti   double or twin room   italy  venice        0.93
2     hotel bartolomeo   double or twin room   italy  venice        0.93
3        hotel moresco   double or twin room   italy  venice        0.93
4         hotel tivoli   double or twin room   italy  venice        0.93
5         hotel falier   double or twin room   italy  venice        0.93
6       hotel malibran  standard double room   italy  venice        0.93
7     hotel bartolomeo  standard double room   italy  venice        0.93
8       hotel malibran  superior double room   italy  venice        0.91
9        hotel moresco  superior double room   italy  venice        0.91

Untuk evaluasi content-based filtering, digunakan precision@k dalam menentukan apakah rekomendasi relevan atau tidak, dengan rumus sebagai berikut:

$$
\text{Precision@K} = \frac{\text{True Positives (TP)}}{\text{True Positives (TP)} + \text{False Positives (FP)}}
$$


Dari hasil rekomendasi:  

| Hotel Name | Room Type | Relevan? |
|------------|--------------|---------|
| Hotel Antica Fenice | Double Room | TP |
| Hotel Dolomiti | Double or Twin Room | TP |
| Hotel Bartolomeo | Double or Twin Room | TP |
| Hotel Moresco | Double or Twin Room | TP |
| Hotel Tivoli | Double or Twin Room | TP |
| Hotel Falier | Double or Twin Room | TP |
| Hotel Malibran | Standard Double Room | TP |
| Hotel Bartolomeo | Standard Double Room | TP |
| Hotel Malibran | Superior Double Room | TP |
| Hotel Moresco | Superior Double Room | TP |

**Perhitungan:**
$$
\text{Precision@10} = \frac{10}{10 + 0} = \frac{10}{10} = 1.0
$$

**Precision@10 untuk Venice adalah 1.0 atau 100%**, yang menunjukkan bahwa rekomendasi sudah baik, dengan 10 dari 10 hasil relevan dengan permintaan pengguna.

In [23]:
recommend('single', 'Spain', 'Barcelona')

hotelname              roomtype country       city  similarity
0        hotel viladomat                single   spain  barcelona        1.00
1     hotel ciutat vella                single   spain  barcelona        1.00
2          astoria hotel                single   spain  barcelona        1.00
3             abba sants                single   spain  barcelona        0.98
4  hotel teatre auditori           single room   spain  barcelona        0.98
5           magatzem 128  interior single room   spain  barcelona        0.96
6             hotel lleo       single standard   spain  barcelona        0.93
7          astoria hotel       standard single   spain  barcelona        0.93
8          moderno hotel       single standard   spain  barcelona        0.93
9       hotel onix liceo      twin single room   spain  barcelona        0.92

$$
\text{Precision@10} = \frac{10}{10 + 0} = \frac{10}{10} = 1.0
$$

**Precision@10 untuk Barcelona adalah 1.0 atau 100%**, yang menunjukkan bahwa rekomendasi sudah baik, dengan 10 dari 10 hasil relevan dengan permintaan pengguna.


### Dapat disimpulkan Fungsi Rekomendasi sudah berjalan sebagaimana mestinya dengan mengembalikan Top-N rekomendasi hotel dengan akurasi presisi relatif di angka 10/10 atau 100%.